lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
+    : _pulledPrims(pulledPrims)
+{
+}
+
+void PrimUpdaterManager::OrphanedNodesManager::operator()(const Ufe::Notification& n)
+{
+    const auto& sceneNotification = static_cast<const Ufe::SceneChanged&>(n);
+    auto        changedPath = sceneNotification.changedPath();
+
+    // No changedPath means composite.  Use containsDescendant(), as
+    // containsDescendantInclusive() would mean a structure change on the
+    // pulled node itself, which is not possible (pulled objects are locked).
+    if (changedPath.empty()) {
+        const auto& sceneCompositeNotification
+            = static_cast<const Ufe::SceneCompositeNotification&>(n);
+        for (const auto& op : sceneCompositeNotification.opsList()) {
+            if (_pulledPrims.containsDescendant(op.path)) {
+                handleOp(op);
+            }
+        }
+    } else if (_pulledPrims.containsDescendant(changedPath)) {
+        handleOp(sceneNotification);
+    }
+}
+
+void PrimUpdaterManager::OrphanedNodesManager::handleOp(
+    const Ufe::SceneCompositeNotification::Op& op)
+{
+    switch (op.opType) {
+    case Ufe::SceneChanged::ObjectAdd: {
+        // Restoring a previously-deleted scene item may restore an orphaned
+        // node.  Traverse the trie, and show hidden pull parents that are
+        // descendants of the argument path.  The trie node that corresponds to
+        // the added path is the starting point.  It may be an internal node,
+        // without data.
+        auto ancestorNode = _pulledPrims.node(op.path);
+        TF_VERIFY(ancestorNode);
+        recursiveSetVisibility(ancestorNode, true);
+    } break;
+    case Ufe::SceneChanged::ObjectDelete: {
+        // The following cases will generate object delete:
+        // - Inactivate of ancestor USD prim sends object post delete.  The
+        //   inactive object has no children.
+        // - Delete of ancestor Maya Dag node, which sends object pre delete.
+        //
+        // At time of writing (25-Aug-2022), delete of an ancestor USD prim
+        // (which sends object destroyed) is prevented by edit restrictions, as
+        // pulling creates an over opinion all along the ancestor chain in the
+        // session layer, which is strongest.  If these restrictions are
+        // lifted, hiding the pull parent is appropriate.
+        //
+        // Traverse the trie, and hide pull parents that are descendants of
+        // the argument path.  First, get the trie node that corresponds to
+        // the path.  It may be an internal node, without data.
+        auto ancestorNode = _pulledPrims.node(op.path);
+        TF_VERIFY(ancestorNode);
+        recursiveSetVisibility(ancestorNode, false);
+    } break;
+    case Ufe::SceneChanged::SubtreeInvalidate: {
+        // On subtree invalidate, the scene item itself has not had a structure
+        // change, but its children have changed.  There are two cases:
+        // - the node has children: from a variant switch, or from a payload
```

Comment:
Ugh, unfortunate that we need to have such intimate knowledge of cause-and-effect of these notifications.